crates/async-dropper-simple/src/lib.rs

```diff
@@ -2,41 +2,98 @@
 //! The code in this file was shamelessly stolen from
 //! https://stackoverflow.com/questions/71541765/rust-async-drop
 
-use std::time::Duration;
+/// Represents an infallible case
+pub static INFALLIBLE: &str = "INFALLIBLE";
+
+use std::{
+    ops::{Deref, DerefMut},
+    time::Duration,
+};
 
 #[async_trait::async_trait]
 pub trait AsyncDrop {
     async fn async_drop(&mut self);
 }
 
-#[derive(Default)]
-#[allow(dead_code)]
-pub struct AsyncDropper<T: AsyncDrop + Default + Send + 'static> {
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct AsyncDropper<T: AsyncDrop + Send> {
     dropped: bool,
     timeout: Option<Duration>,
-    inner: T,
+    inner: Option<T>,
 }
 
-impl<T: AsyncDrop + Default + Send + 'static> AsyncDropper<T> {
+impl<T: AsyncDrop + Send> AsyncDropper<T> {
     pub fn new(inner: T) -> Self {
         Self {
             dropped: false,
             timeout: None,
-            inner,
+            inner: Some(inner),
         }
     }
 
     pub fn with_timeout(timeout: Duration, inner: T) -> Self {
         Self {
             dropped: false,
             timeout: Some(timeout),
-            inner,
+            inner: Some(inner),
+        }
+    }
+
+    /// Get a reference to the inner data.
+    ///
+    /// # Panics
+    ///
+    /// Can never occur since the Drop implementation only `take` the inner once, and only then.
+    pub fn inner(&self) -> &T {
+        self.inner.as_ref().expect(INFALLIBLE)
+    }
+
+    /// Get a mutable reference to the inner data.
+    /// # Panics
+    ///
+    /// Can never occur since the Drop implementation only `take` the inner once, and only then.
+    pub fn inner_mut(&mut self) -> &mut T {
+        self.inner.as_mut().expect(INFALLIBLE)
+    }
+}
+
+/// It doesn't require default, this tradeoff means that you need to use `AsyncDropper::new` or `AsyncDropper::with_timeout`.
+/// Since `AsyncDropper::default()` doesn't create `T::default()` inner, we create a dummy Default instance on `std::mem::take` to be thrown away.
+impl<T> Default for AsyncDropper<T>
+where
+    T: AsyncDrop + Send,
+{
+    fn default() -> Self {
+        Self {
+            dropped: true,
+            timeout: None,
+            inner: None,
         }
     }
 }
 
+impl<T> Deref for AsyncDropper<T>
```

Comment:
Thanks for adding these implementations, they definitely should have been here before! 🙇